src/box/txn_limbo.c

```diff
@@ -553,50 +751,215 @@ txn_limbo_read_rollback(struct txn_limbo *limbo, int64_t lsn)
 	}
 }
 
-int
-txn_limbo_write_promote(struct txn_limbo *limbo, int64_t lsn, uint64_t term)
+static void
+txn_limbo_make_promote(struct txn_limbo *limbo, int64_t lsn, uint64_t term,
+		       struct synchro_request *req)
 {
-	assert(latch_is_locked(&limbo->promote_latch));
 	/*
 	 * We make sure that promote is only written once everything this
 	 * instance has may be confirmed.
 	 */
 	struct txn_limbo_entry *e = txn_limbo_last_synchro_entry(limbo);
 	assert(e == NULL || e->lsn <= lsn);
 	(void) e;
-	struct synchro_request req = {
+
+	/*
+	 * The PROMOTE request we're about to write depends on
+	 * the previous one if it exists. It's an inductive property.
+	 */
+	struct rlist *queue = &limbo->pending_promotes;
+	struct pending_promote *last = !rlist_empty(queue)
+		? rlist_last_entry(queue, typeof(*last), link)
+		: NULL;
+
+	uint64_t greatest_term = limbo->promote_greatest_term;
+	*req = (struct synchro_request) {
 		.type = IPROTO_RAFT_PROMOTE,
-		.replica_id = limbo->owner_id,
 		.origin_id = instance_id,
-		.lsn = lsn,
+		.wait_ack = true,
```

Comment:
You don't seem to need `wait_ack` field. It is always equal to `type == IPROTO_RAFT_PROMOTE`, from what I see. Why did you add it?